### PR TITLE
Support metrics at Source.split for Direct Runner

### PR DESCRIPTION
**Pull request by Abacn**
Address #32021 for direct runner

* Added a validate runner test for this scenario

**Please** add a meaningful description for your change here

Tested manually - BUILD SUCCESSFUL for the following commands

```
./gradlew :runners:direct-java:validatesRunner --tests *testBoundedSourceMetricsInSplit
```

confirmed the new test already passing on portable runners:

```
./gradlew :runners:flink:1.18:job-server:validatesPortableRunnerBatch --tests *testBoundedSourceMetricsInSplit

./gradlew :runners:spark:3:job-server:validatesPortableRunnerBatch --tests *testBoundedSourceMetricsInSplit

./gradlew :runners:google-cloud-dataflow-java:validatesRunnerV2Test -PgcpProject=... -PgcpRegion=us-central1 -PgcpTempRoot=gs://.../tmp --tests *testBoundedSourceMetricsInSplit -info
```

Dataflow legacy runner does not support attempted metrics (and excluded it from validateRunner test). If change to match committed metrics, tested that the test also passes.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

File: sdks/java/core/src/test/java/org/apache/beam/sdk/metrics/MetricResultsMatchers.java [hunk 1]
```diff
@@ -211,17 +212,21 @@ private static class MatchNameAndKey<T> extends TypeSafeMatcher<MetricResult<T>>
 
     private final String namespace;
     private final String name;
-    private final String step;
+    private final @Nullable String step;
```
Comment (Abacn) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/metrics/MetricResultsMatchers.java:215]: `MetricKey.create` has parameter type `(@Nullable String, MetricName)`. It already supported null step

**Comment by Abacn** [comment]
This will make #31804 testable on Direct runner also -- BQ source lineage were added in BoundedSource.split, currently not testable on Direct runner

**Comment by Abacn** [comment]
assign set of reviewers

**Comment by Abacn** [comment]
R: @Naireen @damondouglas

**Review by damondouglas — APPROVED** [review]
@Abacn LGTM

**Pull request merged**